src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/FileWatcherBasedRazorProjectInfoDriver.cs

```diff
@@ -0,0 +1,180 @@
+    {
+        var workspaceDirectoryPath = await _workspaceRootPathProvider.GetRootPathAsync(cancellationToken).ConfigureAwait(false);
+        workspaceDirectoryPath = FilePathNormalizer.Normalize(workspaceDirectoryPath);
+
+        var existingConfigurationFiles = DirectoryHelper.GetFilteredFiles(
+            workspaceDirectoryPath,
+            _options.ProjectConfigurationFileName,
+            s_ignoredDirectories,
+            logger: _logger).ToImmutableArray();
+
+        foreach (var filePath in existingConfigurationFiles)
+        {
+            using var stream = File.OpenRead(filePath);
+
+            var razorProjectInfo = await TryDeserializeAsync(filePath, cancellationToken).ConfigureAwait(false);
+
+            if (razorProjectInfo is not null)
+            {
+                EnqueueUpdate(razorProjectInfo);
+            }
+        }
+
+        if (!Directory.Exists(workspaceDirectoryPath))
+        {
+            _logger.LogInformation($"Creating workspace directory: '{workspaceDirectoryPath}'");
+            Directory.CreateDirectory(workspaceDirectoryPath);
+        }
+
+        _logger.LogInformation($"Starting {nameof(FileWatcherBasedRazorProjectInfoDriver)}: '{workspaceDirectoryPath}'");
+
+        _fileSystemWatcher = new RazorFileSystemWatcher(workspaceDirectoryPath, _options.ProjectConfigurationFileName)
+        {
+            NotifyFilter = NotifyFilters.FileName | NotifyFilters.LastWrite | NotifyFilters.CreationTime,
+            IncludeSubdirectories = true,
+        };
+
+        _fileSystemWatcher.Created += (_, args) => EnqueueAddOrChange(args.FullPath);
+        _fileSystemWatcher.Changed += (_, args) => EnqueueAddOrChange(args.FullPath);
+        _fileSystemWatcher.Deleted += (_, args) => EnqueueRemove(args.FullPath);
+        _fileSystemWatcher.Renamed += (_, args) => EnqueueRename(args.OldFullPath, args.FullPath);
+
+        _fileSystemWatcher.EnableRaisingEvents = true;
+
+        void EnqueueAddOrChange(string filePath)
+        {
+            _workQueue.AddWork((filePath, ChangeKind.AddOrUpdate));
+        }
+
+        void EnqueueRemove(string filePath)
+        {
+            _workQueue.AddWork((filePath, ChangeKind.Remove));
+        }
+
+        void EnqueueRename(string oldFilePath, string newFilePath)
+        {
+            EnqueueRemove(oldFilePath);
+            EnqueueAddOrChange(newFilePath);
+        }
+    }
+
+    private async ValueTask ProcessBatchAsync(ImmutableArray<(string FilePath, ChangeKind Kind)> items, CancellationToken token)
+    {
+        foreach (var (filePath, changeKind) in items.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (token.IsCancellationRequested)
+            {
+                return;
+            }
+
+            switch (changeKind)
+            {
+                case ChangeKind.AddOrUpdate:
+                    using (var stream = File.OpenRead(filePath))
```

Comment:
Did you mean for `TryDeserialize` to take in the stream? or can this block be removed?

Reply:
Yes, thank you! I missed that when refactoring a bit.